### PR TITLE
Add pre-commit action

### PR DESCRIPTION
**Pull request by dstansby**

### CODE REVIEW

File: pre-commit/action.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+          ${{ env.pythonLocation }}
+          ~/.cache/pre-commit
+        key: |
+          pre-commit-${{ env.PY }}-${{ hashFiles('.pre-commit-config.yaml') }}
+
+    - name: Run pre-commit
+      shell: bash
+      run: |
+        python -m pip install pre-commit --upgrade
+        pre-commit run --all-files --color always
```
Comment (paddyroddy) [inline pre-commit/action.yml]: This won't currently pass `end-of-file-fixer` above 😉

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+repos:
+-   repo: https://github.com/pre-commit/pre-commit-hooks
```
Comment (paddyroddy) [inline .pre-commit-config.yaml:2]: I'm assuming this file is just a dummy example?

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+    rev: v3.2.0
+    hooks:
+    -   id: trailing-whitespace
+    -   id: end-of-file-fixer
+    -   id: check-yaml
+    -   id: check-added-large-files
```
Comment (paddyroddy) [inline .pre-commit-config.yaml:8]: From the above, other useful generic ones (full list [here](https://github.com/pre-commit/pre-commit-hooks)):
* `check-case-conflict`
* `check-executables-have-shebangs`
* `check-merge-conflict`
* `mixed-line-ending` with `args: [--fix=lf]` for Windows users

**Review by paddyroddy — APPROVED** [review]

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+repos:
+-   repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v3.2.0
```
Comment (paddyroddy) [inline .pre-commit-config.yaml:3]:
```suggestion
    rev: v4.1.0
```

File: .pre-commit-config.yaml (hunk 2 above)
Comment (dstansby) [reply .pre-commit-config.yaml:2]: Yep, but it is run/applied to the repo.

File: .pre-commit-config.yaml (hunk 4 above)
Comment (dstansby) [reply .pre-commit-config.yaml:3]: I enabled dependabot for this repo, so it should take care of opening PRs for verison updates

Comment (paddyroddy) [reply .pre-commit-config.yaml:3]: Dependabot doesn't support `pre-commit` https://github.com/dependabot/dependabot-core/issues/1524

Comment (dstansby) [reply .pre-commit-config.yaml:3]: Oh woops, I thought this was a github actions file (not sure why). Is there something else I can use to update all the versions locally, or do I have to go through each one manually?

Comment (paddyroddy) [reply .pre-commit-config.yaml:3]: You can run `pre-commit autoupdate`. Not sure of a way of automating it. I just run that command every now and then.

Comment (dstansby) [reply .pre-commit-config.yaml:3]: I found `pre-commit autoupdate` 😄

**Pull request merged**